### PR TITLE
Update testing documentation in readme

### PR DESCRIPTION
**Pull request by jone**
I have updated the testing documentation in the README:

- Remove the old gever-testbrowser documentation; `ftw.testbrowser` shall be used.
- Add a section about functional and integration testing.
- Add an example integration test case.
- Add some best practices.

@4teamwork/gever please have a look.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -365,63 +368,221 @@ The ``bin/mtest`` script can be configured with environment variables:
...
+``Manager``-user, the ``plone.app.testing``'s site owner may be used.
+
+Loggin users.
+^^^^^^^^^^^^^
+
+Integration testing tests start with a low-level security.
+Use the ``self.login`` method for setting up the needed security within the
+test code.
+
+TODO: Document browser login behavior https://4teamwork.slack.com/archives/C060FEYKA/p1498490241485990
```
Comment (jone) [inline README.rst]: 🚧

File: README.rst [hunk 2]
```diff
@@ -365,63 +368,221 @@ The ``bin/mtest`` script can be configured with environment variables:
...
 
-The center of the `Browser API` is the ``OGBrowser`` class. It's a
-simple subclass of ``plone.testing.z2.Browser`` and the easiest way to
-use it is to extend ``opengever.testing.FunctionalTestCase``:
+Use ``tearDown`` carefully.
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Do not tear down changes which are taken care of by some kind of isolation:
+
+- Do *not* tear down ZODB changes: the ZDOB is isolated by ``plone.app.testing``.
```
Comment (lukasgraf) [inline README.rst]: `ZDOB` -> `ZODB`

File: README.rst [hunk 3]
```diff
@@ -365,63 +368,221 @@ The ``bin/mtest`` script can be configured with environment variables:
...
+- Do *not* tear down SQL changes: we take care of that in the SQL testing layer
+  with savepoints / rollbacks.
+- Do *not* tear down component registry changes (e.g. new adapters, utilities,
+  event handlers) as this is taken care of by the
+  `COMPONENT_REGISTRY_ISOLATION`_ layer.
+- *Do* tear down modifications in evnironment variables (``os.environ``).
```
Comment (lukasgraf) [inline README.rst]: `evnironment` -> `environment`

File: README.rst [hunk 4]
```diff
@@ -365,63 +368,221 @@ The ``bin/mtest`` script can be configured with environment variables:
...
+.. code:: python
+
+    def test_closing_dossier(self):
+        self.assertTrue(self.dossier.is_open(),
+                        'Precondition: assumed dossier to be open')
+        self.dossier.close()
+        self.assertFalse(self.dossier.is_open())
+
+If the ``self.dossier`` is changed to be not open by default anymore, the failure
+should tell us that a procondition was no longer met rather than the ``close()``
```
Comment (lukasgraf) [inline README.rst]: `procondition` -> `precondition`, `rather than` -> `rather than implying that`

File: README.rst [hunk 5]
```diff
@@ -365,63 +368,221 @@ The ``bin/mtest`` script can be configured with environment variables:
...
+
+.. code:: python
+
+    def test_title_is_journalized_on_action(self):
+        self.dossier.title = u'The dossier'
+        action(self.dossier)
+        self.assertEquals(u'The dossier',
+                          last_journal_entry(self.dossier).title)
+
+Activating feature flags.
```
Comment (lukasgraf) [inline README.rst]: I would leave the period at the end of those sections off (most of them are not full sentences). </nitpick>

File: README.rst [hunk 6]
```diff
@@ -365,63 +368,235 @@ The ``bin/mtest`` script can be configured with environment variables:
...
+The fixture provides a set of standard users which should be used in tests.
+Do not use ``plone.app.testing``'s test user with global roles as it does
+not reflect properly how the security model of GEVER works.
+In order to test features which can only be executed by the system or by a
+``Manager``-user, the ``plone.app.testing``'s site owner may be used.
+
+Login
+^^^^^
+
+Integration testing tests start with a low-level security.
```
Comment (jone) [inline README.rst]: > Im not 100% sure what you mean by this sentence. "Security" as in Security Manager as in.. User? I.e. `low-level security == user with minimal privileges`? Doesn't make sense to me though, because by default (without `login()`) it would be `Anonymous`, right?

Current implementation:
- in the test-code, the security manager is authorized as "self.reguler_user": a user who simply is in the users group of the client, no more responsibilities (see [test case setUp](https://github.com/4teamwork/opengever.core/blob/master/opengever/testing/integration_test_case.py#L25) and [fixtures](https://github.com/4teamwork/opengever.core/blob/master/opengever/testing/fixtures.py#L66-L67)),
- but the browser is not logged in at all (anonymous), as it is default in ``ftw.testbrowser``

I would like to change the behavior so that no user is logged in at all when a tests begins, because:
- as a test reader, I want to see for each test with which user it runs (as in BBD `As a dossier responsible, In order to .., I want to ..`)
- from a security / QA perspective I want to lead devs to making an active deicision for who this test should be
- as a developer I would like to have consistency between test-code and browser-code regarding who is the user under test.

🚧 I will come up with a PR separately which changes the code

Comment (lukasgraf) [inline README.rst]: Ah, I see, thanks for clarifying! 👍  (and I agree with the proposed future changes).
I would rephrase that paragraph to something like that then:

> Integration tests start with logging in / assuming the security context of a user with minimal privileges (`self.regular_user`) on the server side. On the client side (browser), no user is logged in by default (Anonymous).
>
> Use the ``self.login`` method for setting up the needed security context within the test code.
[...]
```

Comment (jone) [inline README.rst]: Ok, I like your suggestion and I've amended it.
I'll work on changing the defaults as explained above and then change this part again in order to make this PR mergeable already.

File: README.rst [hunk 7]
```diff
@@ -365,63 +368,235 @@ The ``bin/mtest`` script can be configured with environment variables:
...
+   from ftw.testbrowser import browsing
+   from ftw.testbrowser.pages import statusmessages
+   from opengever.testing import IntegrationTestCase
 
+   class TestExampleView(IntegrationTestCase):
 
-Then you can use the ``Builder`` function in your test cases:
+       @browsing
+       def test_example_view(self, browser):
+           self.login(self.dossier_responsible)
```
Comment (deiferni) [inline README.rst]: 😄 
```python
self.login(self.dossier_responsible, browser)
```

File: README.rst [hunk 8]
```diff
@@ -365,63 +368,235 @@ The ``bin/mtest`` script can be configured with environment variables:
...
+reuse objects from the fixture and modify them as needed.
+
+Use users from fixture
+^^^^^^^^^^^^^^^^^^^^^^
+
+The fixture provides a set of standard users which should be used in tests.
+Do not use ``plone.app.testing``'s test user with global roles as it does
+not reflect properly how the security model of GEVER works.
+In order to test features which can only be executed by the system or by a
+``Manager``-user, the ``plone.app.testing``'s site owner may be used.
```
Comment (deiferni) [inline README.rst]: should we alias the site-owner as self.manager? should we actively prevent that the discouraged test-user is used?

File: README.rst (hunk 7 above)
Comment (jone) [inline README.rst]: @deiferni The feature, which logs in the browser here, is not yet implemented. I will amend the documentation in the PR implementing the feature.

File: README.rst (hunk 8 above)
Comment (jone) [inline README.rst]: > should we alias the site-owner as self.manager?
I can do that, but I'd like to do that in a separate PR. 😎 

> should we actively prevent that the discouraged test-user is used?
The user has not enough priveleges to do anything by default and `self.grant` is not implemented, so the user is not really usable.
If a dev is still really serious about using test user, we should see and handle that in the review.
I therefore think that we should not spend more time on building guards regarding that 😉

File: README.rst (hunk 7 above)
Comment (jone) [inline README.rst]: @jone this is wrong, it is actually already implemented in `master`.
But it is corrected anyway in the next PR..

Comment (jone) [inline README.rst]: => #3072

File: README.rst (hunk 8 above)
Comment (jone) [inline README.rst]: => #3073
